main.go

```diff
@@ -49,75 +39,70 @@ func failf(format string, v ...interface{}) {
-	if err != nil {
-		return false, formatErr(dev, err)
-	}
+func getBootCompleteEvent(adbManager *adbmanager.Model, serial string, timeout time.Duration) <-chan WaitForBootCompleteResult {
+	doneChan := make(chan WaitForBootCompleteResult)
 
-	sys, err := cmdRunner.RunCommandWithTimeout(adbShellCommand(androidHome, serial, "getprop sys.boot_completed"))
-	if err != nil {
-		return false, formatErr(sys, err)
-	}
+	go func() {
+		time.AfterFunc(timeout, func() {
+			doneChan <- WaitForBootCompleteResult{Error: errors.New("timeout")}
+		})
+	}()
 
-	init, err := cmdRunner.RunCommandWithTimeout(adbShellCommand(androidHome, serial, "getprop init.svc.bootanim"))
-	if err != nil {
-		return false, formatErr(init, err)
-	}
+	go func() {
+		out, err := adbManager.WaitForDeviceThenShellCmd(serial, nil, "getprop sys.boot_completed").RunAndReturnTrimmedCombinedOutput()
+		if err != nil {
+			fmt.Println(out)
+			doneChan <- WaitForBootCompleteResult{Error: err}
+			return
+		}
 
-	return dev == "1" && sys == "1" && init == "stopped", nil
-}
+		lines := strings.Split(out, "\n")
+		for _, line := range lines {
+			if strings.TrimSpace(line) == "1" {
+				doneChan <- WaitForBootCompleteResult{Booted: true}
+				return
+			}
+		}
+
+		doneChan <- WaitForBootCompleteResult{Booted: false}
+	}()
 
-func terminateADBServer(androidHome string) error {
-	name, args := adbCommand(androidHome, "", "kill-server")
-	_, err := cmdRunner.RunCommandWithTimeout(name, args)
-	return err
+	return doneChan
 }
 
-func checkEmulatorBootState(androidHome, emulatorSerial string, timeout time.Duration) error {
+func waitForDevice(adb *adbmanager.Model, emulatorSerial string, timeout time.Duration) error {
 	startTime := clock.Now()
 
 	for {
-		logger.Printf("> Checking if boot completed...")
-
-		booted, err := isDeviceBooted(androidHome, emulatorSerial)
-		if err != nil {
-			switch {
-			case strings.Contains(err.Error(), "daemon not running; starting now at"):
-				logger.Warnf("adb daemon being restarted")
-				logger.Printf(err.Error())
-			case err == errTimedOut:
-				logger.Warnf("Running command timed out, retry...")
-				if err := terminateADBServer(androidHome); err != nil {
-					if err != errTimedOut {
-						return fmt.Errorf("unable to kill ADB daemon, error: %s", err)
-					}
-					logger.Warnf("killing ADB daemon timed out")
-				}
-			case err != nil:
-				return fmt.Errorf("failed to check emulator boot status: %s", err)
+		logger.Printf("Waiting for emulator boot...")
+
+		bootCompleteChan := getBootCompleteEvent(adb, emulatorSerial, timeout)
+		result := <-bootCompleteChan
+		switch {
+		case result.Error != nil:
+			logger.Warnf("Failed to check emulator boot status: %s", result.Error)
+			logger.Warnf("Killing ADB server before retry...")
+			if err := adb.KillServerCmd(nil).Run(); err != nil {
+				return fmt.Errorf("failed to terminate adb server: %s", err)
```

Comment:
probably error will be an `exist status X`, which is not too descriptive. I suggest including the cmd output too.